proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,566 @@
+
+### Syscall to change PRAW fees
+
+PRAW fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+A new syscall will be added to update the PRAW fees:
+
+```
+fn update_program_rebatable_account_write_fees(
+  account: Pubkey, 
+  milli_lamports_per_cu: u32
+) -> bool;
+```
+
+The syscall will take as arguments the account on which fee should be updated
+and the millilamports per cu that should be charged and return a true if the
+changes were successful else will return false.
+
+The accounts database will need to store the value of millilamports per
+requested CU for the account. Changes to update the PRAW fees could be done only
+by the authority program. All programs will need to implement special
+instructions so that the program admin can use this syscall to update PRAW fees
+on the required accounts. Program admin should always be signer for these
+instructions.
+
+Setting the millilamports per CU to `0` disables PRAW fees. The maximum amount
+of millilamports per CU that could be set will be u32::MAX. Lets assume that the
+maximum CU required to update these fees on an account is 1,000 CU. So the
+maximum PRAW fees required to update an account (`M`) is
+
+```
+M = u32::MAX * 1000 CU * (1/1000) lamports
+M = 4.294967296×10^9 * 1000 / 1000
+M = 4.294967296 SOLs
+```
+
+So this makes sure that admin does not lock itself out from changing the PRAW
+fees.
+
+### Syscall to read PRAW fees
+
+A new syscall will be added to read PRAW fees on an account.
+
+```
+fn get_program_rebatable_account_write_fees(
+  account: Pubkey
+) -> u32;
+```
+
+The syscall will take writable account as import and return millilamports per CU.
+
+### Syscall to Rebate
+
+The owner of the account can issue a rebate of the PRAW fees paid to write lock
+a specific account. Similar to the update and read PRAW fees this will need to
+be exposed through a syscall so that programs can implement custom authorization
+and delegate the decision to composing programs.
+
+Simple rebate schemes will verify merely signers, e.g an oracle preventing 3rd
+parties from write-locking their price feed. Dexes will need more complex rebate
+schemes based on instruction sysvar introspection.
+
+Syscall definition of the rebate mechanism is:
+
+```
+fn rebate_program_rebatable_account_write_fees(
+  account: Pubkey, 
+  milli_lamports_per_cu: u32
+) -> bool;
+```
+
+Rebate takes account on which rebate is issued and the amount of millilamports
+per CU that needs to be rebated as input. It returns true of rebate was
+successful false otherwise. In case of multiple rebates on the same account only
+the highest amount of rebate will be taken into account. The rebated amount is
+always the minimum of rebate issued by the program and the PRAW fees on the
+account. If program rebates `U64::MAX` it means all the PRAW fees on the account
```

Comment:
This suggestion is still slightly clunky, maybe it can be reworked better, but it took me a while to realize what was actually going on here:
```suggestion
the highest amount of rebate will be taken into account. The rebated amount is
always the minimum of the largest call to `rebate_program_rebatable_account_write_fees()` and the PRAW fees on the
account. If program rebates `U64::MAX` it means all the PRAW fees on the account
```